/*
 * @file    TestReadFromFileL3V2_2.java
 * @brief   Reads test-data/l3v2-empty-lo-1.xml into memory and tests it.
 *
 * @author  Akiya Jouraku (Java conversion)
 * @author  Sarah Keating 
 * 
 * ====== WARNING ===== WARNING ===== WARNING ===== WARNING ===== WARNING ======
 *
 * DO NOT EDIT THIS FILE.
 *
 * This file was generated automatically by converting the file located at
 * src/sbml/test/TestReadFromFileL3V2_2.cpp
 * using the conversion program dev/utilities/translateTests/translateTests.pl.
 * Any changes made here will be lost the next time the file is regenerated.
 *
 * -----------------------------------------------------------------------------
 * This file is part of libSBML.  Please visit http://sbml.org for more
 * information about SBML, and the latest version of libSBML.
 *
 * Copyright 2005-2010 California Institute of Technology.
 * Copyright 2002-2005 California Institute of Technology and
 *                     Japan Science and Technology Corporation.
 * 
 * This library is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation.  A copy of the license agreement is provided
 * in the file named "LICENSE.txt" included with this software distribution
 * and also available online as http://sbml.org/software/libsbml/license.html
 * -----------------------------------------------------------------------------
 */

package org.sbml.libsbml.test.sbml;

import org.sbml.libsbml.*;

import java.io.File;
import java.lang.AssertionError;

public class TestReadFromFileL3V2_2 {

  static void assertTrue(boolean condition) throws AssertionError
  {
    if (condition == true)
    {
      return;
    }
    throw new AssertionError();
  }

  static void assertEquals(Object a, Object b) throws AssertionError
  {
    if ( (a == null) && (b == null) )
    {
      return;
    }
    else if ( (a == null) || (b == null) )
    {
      throw new AssertionError();
    }
    else if (a.equals(b))
    {
      return;
    }

    throw new AssertionError();
  }

  static void assertNotEquals(Object a, Object b) throws AssertionError
  {
    if ( (a == null) && (b == null) )
    {
      throw new AssertionError();
    }
    else if ( (a == null) || (b == null) )
    {
      return;
    }
    else if (a.equals(b))
    {
      throw new AssertionError();
    }
  }

  static void assertEquals(boolean a, boolean b) throws AssertionError
  {
    if ( a == b )
    {
      return;
    }
    throw new AssertionError();
  }

  static void assertNotEquals(boolean a, boolean b) throws AssertionError
  {
    if ( a != b )
    {
      return;
    }
    throw new AssertionError();
  }

  static void assertEquals(int a, int b) throws AssertionError
  {
    if ( a == b )
    {
      return;
    }
    throw new AssertionError();
  }

  static void assertNotEquals(int a, int b) throws AssertionError
  {
    if ( a != b )
    {
      return;
    }
    throw new AssertionError();
  }

  public void test_echo_l3v2_empty_lo_1()
  {
    String expected = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n" + 
    "<sbml xmlns=\"http://www.sbml.org/sbml/level3/version2/core\" " + 
    "level=\"3\" version=\"2\">\n"  + 
    "  <model substanceUnits=\"mole\" timeUnits=\"second\" volumeUnits=\"litre\" " + 
    "areaUnits=\"metre\" lengthUnits=\"metre\" extentUnits=\"mole\" " + 
    "conversionFactor=\"p\">\n" + 
    "    <listOfFunctionDefinitions id=\"s\"/>\n" + 
    "    <listOfUnitDefinitions name=\"ss\"/>\n" + 
    "    <listOfCompartments id=\"s1\"/>\n" + 
    "    <listOfSpecies>\n" + 
    "      <annotation>\n" + 
    "        <rdf:RDF xmlns:rdf=\"http://www.w3.org/1999/02/22-rdf-syntax-ns#\" " + 
    "xmlns:dc=\"http://purl.org/dc/elements/1.1/\" " + 
    "xmlns:dcterms=\"http://purl.org/dc/terms/\" " + 
    "xmlns:vCard=\"http://www.w3.org/2001/vcard-rdf/3.0#\" " + 
    "xmlns:bqbiol=\"http://biomodels.net/biology-qualifiers/\" " + 
    "xmlns:bqmodel=\"http://biomodels.net/model-qualifiers/\">\n" + 
    "          <rdf:Description rdf:about=\"#_000002\">\n" + 
    "            <bqbiol:is>\n" + 
    "              <rdf:Bag>\n" + 
    "                <rdf:li rdf:resource=\"http://www.geneontology.org/#GO:0007274\"/>\n" + 
    "                <rdf:li rdf:resource=\"http://www.geneontology.org/#GO:0015464\"/>\n" + 
    "                <rdf:li rdf:resource=\"http://www.genome.jp/kegg/pathway/#hsa04080\"/>\n" + 
    "                <rdf:li rdf:resource=\"http://www.ncbi.nlm.nih.gov/Taxonomy/#7787\"/>\n" + 
    "              </rdf:Bag>\n" + 
    "            </bqbiol:is>\n" + 
    "          </rdf:Description>\n" + 
    "        </rdf:RDF>\n" + 
    "      </annotation>\n" + 
    "    </listOfSpecies>\n" + 
    "    <listOfParameters id=\"d\"/>\n" + 
    "    <listOfInitialAssignments name=\"55\"/>\n" + 
    "    <listOfRules>\n" + 
    "      <notes>\n" + 
    "        <body xmlns=\"http://www.w3.org/1999/xhtml\">\n" + 
    "          <center>\n" + 
    "            <h2>A Simple Mitotic Oscillator</h2>\n" + 
    "          </center>\n" + 
    "          <p>A minimal cascade model for the mitotic oscillator involving cyclin and cdc2 kinase</p>\n" + 
    "        </body>\n" + 
    "      </notes>\n" + 
    "    </listOfRules>\n" + 
    "    <listOfConstraints id=\"c\"/>\n" + 
    "    <listOfReactions name=\"v\"/>\n" + 
    "    <listOfEvents id=\"s\"/>\n" + 
    "  </model>\n" + 
    "</sbml>\n";
    String filename = new String( "../../sbml/sbml/test/test-data/" );
    filename += "l3v2-empty-lo-1.xml";
    String fileout = new String( "../../sbml/sbml/test/test-data/" );
    fileout += "tempOut.xml";
    SBMLDocument d = libsbml.readSBML();
    if (d == null);
    {
    }
    libsbml.writeSBML(d);
    SBMLDocument outD = libsbml.readSBML();
    Model m = outD.getModel();
    assertTrue( m != null );
    assertTrue( m.getNumUnitDefinitions() == 0 );
    assertTrue( m.getListOfUnitDefinitions().isSetName() == true );
    assertTrue( m.getListOfUnitDefinitions().getName().equals( "ss") );
    assertTrue( m.getNumFunctionDefinitions() == 0 );
    assertTrue( m.getListOfFunctionDefinitions().isSetId() == true );
    assertTrue( m.getListOfFunctionDefinitions().getId().equals( "s") );
    assertTrue( m.getNumCompartments() == 0 );
    assertTrue( m.getListOfCompartments().isSetId() == true );
    assertTrue( m.getListOfCompartments().getId().equals( "s1") );
    assertTrue( m.getNumSpecies() == 0 );
    assertTrue( m.getListOfSpecies().isSetAnnotation() == true );
    assertTrue( m.getListOfSpecies().isSetNotes() == false );
    assertTrue( m.getNumParameters() == 0 );
    assertTrue( m.getListOfParameters().isSetId() == true );
    assertTrue( m.getListOfParameters().getId().equals( "d") );
    assertTrue( m.getNumRules() == 0 );
    assertTrue( m.getListOfRules().isSetAnnotation() == false );
    assertTrue( m.getListOfRules().isSetNotes() == true );
    assertTrue( m.getNumInitialAssignments() == 0 );
    assertTrue( m.getListOfInitialAssignments().isSetName() == true );
    assertTrue( m.getListOfInitialAssignments().getName().equals( "55") );
    assertTrue( m.getNumConstraints() == 0 );
    assertTrue( m.getListOfConstraints().isSetName() == false );
    assertTrue( m.getListOfConstraints().isSetId() == true );
    assertTrue( m.getListOfConstraints().getId().equals( "c") );
    assertTrue( m.getNumReactions() == 0 );
    assertTrue( m.getListOfReactions().isSetName() == true );
    assertTrue( m.getListOfReactions().isSetId() == false );
    assertTrue( m.getListOfReactions().getName().equals( "v") );
    assertTrue( m.getNumEvents() == 0 );
    assertTrue( m.getListOfEvents().isSetName() == false );
    assertTrue( m.getListOfEvents().isSetId() == true );
    assertTrue( m.getListOfEvents().getId().equals( "s") );
    String S = libsbml.writeSBMLToString(outD);
    assertEquals( true, equals(expected,S) );
    d = null;
    outD = null;
  }

  public void test_echo_l3v2_empty_lo_3()
  {
    String expected = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n" + 
    "<sbml xmlns=\"http://www.sbml.org/sbml/level3/version2/core\" " + 
    "level=\"3\" version=\"2\">\n"  + 
    "  <model substanceUnits=\"mole\" timeUnits=\"second\" volumeUnits=\"litre\" " + 
    "areaUnits=\"metre\" lengthUnits=\"metre\" extentUnits=\"mole\" " + 
    "conversionFactor=\"p\">\n" + 
    "    <listOfUnitDefinitions>\n" + 
    "      <unitDefinition id=\"nonsense\">\n" + 
    "        <listOfUnits/>\n" + 
    "      </unitDefinition>\n" + 
    "    </listOfUnitDefinitions>\n" + 
    "    <listOfReactions>\n" + 
    "      <reaction id=\"in\" reversible=\"false\" compartment=\"comp\">\n" + 
    "        <listOfReactants/>\n" + 
    "        <listOfProducts name=\"ss\"/>\n" + 
    "        <listOfModifiers sboTerm=\"SBO:0000002\"/>\n" + 
    "        <kineticLaw>\n" + 
    "          <math xmlns=\"http://www.w3.org/1998/Math/MathML\">\n" + 
    "            <apply>\n" + 
    "              <times/>\n" + 
    "              <ci> k1 </ci>\n" + 
    "              <ci> X0 </ci>\n" + 
    "              <ci> cell </ci>\n" + 
    "            </apply>\n" + 
    "          </math>\n" + 
    "          <listOfLocalParameters>\n" + 
    "            <notes>\n" + 
    "              <body xmlns=\"http://www.w3.org/1999/xhtml\">\n" + 
    "                <center>\n" + 
    "                  <h2>A Simple Mitotic Oscillator</h2>\n" + 
    "                </center>\n" + 
    "                <p>A minimal cascade model for the mitotic oscillator involving cyclin and cdc2 kinase</p>\n" + 
    "              </body>\n" + 
    "            </notes>\n" + 
    "          </listOfLocalParameters>\n" + 
    "        </kineticLaw>\n" + 
    "      </reaction>\n" + 
    "    </listOfReactions>\n" + 
    "    <listOfEvents/>\n" + 
    "  </model>\n" + 
    "</sbml>\n";
    String filename = new String( "../../sbml/sbml/test/test-data/" );
    filename += "l3v2-empty-lo-3.xml";
    String fileout = new String( "../../sbml/sbml/test/test-data/" );
    fileout += "tempOut.xml";
    SBMLDocument d = libsbml.readSBML();
    if (d == null);
    {
    }
    libsbml.writeSBML(d);
    SBMLDocument outD = libsbml.readSBML();
    Model m = outD.getModel();
    assertTrue( m != null );
    assertTrue( m.getNumUnitDefinitions() == 1 );
    assertTrue( m.getNumFunctionDefinitions() == 0 );
    assertTrue( m.getNumCompartments() == 0 );
    assertTrue( m.getNumSpecies() == 0 );
    assertTrue( m.getNumParameters() == 0 );
    assertTrue( m.getNumRules() == 0 );
    assertTrue( m.getNumInitialAssignments() == 0 );
    assertTrue( m.getNumConstraints() == 0 );
    assertTrue( m.getNumReactions() == 1 );
    assertTrue( m.getNumEvents() == 0 );
    UnitDefinition ud = m.getUnitDefinition(0);
    assertTrue( ud != null );
    assertTrue( ud.getNumUnits() == 0 );
    Reaction r = m.getReaction(0);
    assertTrue( r != null );
    assertTrue( r.getNumReactants() == 0 );
    assertTrue( r.getNumProducts() == 0 );
    assertTrue( r.getListOfProducts().isSetName() == true );
    assertTrue( r.getListOfProducts().getName().equals( "ss") );
    assertTrue( r.getNumModifiers() == 0 );
    assertTrue( r.getListOfModifiers().isSetSBOTerm() == true );
    assertTrue( r.getListOfModifiers().getSBOTerm() == 2 );
    KineticLaw kl = r.getKineticLaw();
    assertTrue( kl != null );
    assertTrue( kl.getNumLocalParameters() == 0 );
    assertTrue( kl.getListOfLocalParameters().isSetNotes() == true );
    assertTrue( kl.getListOfLocalParameters().isSetAnnotation() == false );
    String S = libsbml.writeSBMLToString(outD);
    assertEquals( true, equals(expected,S) );
    d = null;
    outD = null;
  }

  public void test_read_l3v2_empty_lo_1()
  {
    SBMLReader reader = new SBMLReader();
    SBMLDocument d;
    Model m;
    String filename = new String( "../../sbml/sbml/test/test-data/" );
    filename += "l3v2-empty-lo-1.xml";
    d = reader.readSBML(filename);
    if (d == null);
    {
    }
    assertTrue( d.getLevel() == 3 );
    assertTrue( d.getVersion() == 2 );
    m = d.getModel();
    assertTrue( m != null );
    assertTrue( m.getNumUnitDefinitions() == 0 );
    assertTrue( m.getListOfUnitDefinitions().isSetName() == true );
    assertTrue( m.getListOfUnitDefinitions().getName().equals( "ss") );
    assertTrue( m.getNumFunctionDefinitions() == 0 );
    assertTrue( m.getListOfFunctionDefinitions().isSetId() == true );
    assertTrue( m.getListOfFunctionDefinitions().getId().equals( "s") );
    assertTrue( m.getNumCompartments() == 0 );
    assertTrue( m.getListOfCompartments().isSetId() == true );
    assertTrue( m.getListOfCompartments().getId().equals( "s1") );
    assertTrue( m.getNumSpecies() == 0 );
    assertTrue( m.getListOfSpecies().isSetAnnotation() == true );
    assertTrue( m.getListOfSpecies().isSetNotes() == false );
    assertTrue( m.getNumParameters() == 0 );
    assertTrue( m.getListOfParameters().isSetId() == true );
    assertTrue( m.getListOfParameters().getId().equals( "d") );
    assertTrue( m.getNumRules() == 0 );
    assertTrue( m.getListOfRules().isSetAnnotation() == false );
    assertTrue( m.getListOfRules().isSetNotes() == true );
    assertTrue( m.getNumInitialAssignments() == 0 );
    assertTrue( m.getListOfInitialAssignments().isSetName() == true );
    assertTrue( m.getListOfInitialAssignments().getName().equals( "55") );
    assertTrue( m.getNumConstraints() == 0 );
    assertTrue( m.getListOfConstraints().isSetName() == false );
    assertTrue( m.getListOfConstraints().isSetId() == true );
    assertTrue( m.getListOfConstraints().getId().equals( "c") );
    assertTrue( m.getNumReactions() == 0 );
    assertTrue( m.getListOfReactions().isSetName() == true );
    assertTrue( m.getListOfReactions().isSetId() == false );
    assertTrue( m.getListOfReactions().getName().equals( "v") );
    assertTrue( m.getNumEvents() == 0 );
    assertTrue( m.getListOfEvents().isSetName() == false );
    assertTrue( m.getListOfEvents().isSetId() == true );
    assertTrue( m.getListOfEvents().getId().equals( "s") );
    d = null;
  }

  /**
   * Loads the SWIG-generated libSBML Java module when this class is
   * loaded, or reports a sensible diagnostic message about why it failed.
   */
  static
  {
    String varname;
    String shlibname;

    if (System.getProperty("mrj.version") != null)
    {
      varname = "DYLD_LIBRARY_PATH";    // We're on a Mac.
      shlibname = "libsbmlj.jnilib and/or libsbml.dylib";
    }
    else
    {
      varname = "LD_LIBRARY_PATH";      // We're not on a Mac.
      shlibname = "libsbmlj.so and/or libsbml.so";
    }

    try
    {
      System.loadLibrary("sbmlj");
      // For extra safety, check that the jar file is in the classpath.
      Class.forName("org.sbml.libsbml.libsbml");
    }
    catch (SecurityException e)
    {
      e.printStackTrace();
      System.err.println("Could not load the libSBML library files due to a"+
                         " security exception.\n");
      System.exit(1);
    }
    catch (UnsatisfiedLinkError e)
    {
      e.printStackTrace();
      System.err.println("Error: could not link with the libSBML library files."+
                         " It is likely\nyour " + varname +
                         " environment variable does not include the directories\n"+
                         "containing the " + shlibname + " library files.\n");
      System.exit(1);
    }
    catch (ClassNotFoundException e)
    {
      e.printStackTrace();
      System.err.println("Error: unable to load the file libsbmlj.jar."+
                         " It is likely\nyour -classpath option and CLASSPATH" +
                         " environment variable\n"+
                         "do not include the path to libsbmlj.jar.\n");
      System.exit(1);
    }
  }
}
